WordPress/Sniffs/NamingConventions/ValidHookNameSniff.php

```diff
@@ -0,0 +1,289 @@
+<?php
+/**
+ * WordPress Coding Standard.
+ *
+ * @category PHP
+ * @package  PHP_CodeSniffer
+ * @link     https://make.wordpress.org/core/handbook/best-practices/coding-standards/
+ */
+
+/**
+ * Use lowercase letters in action and filter names. Separate words via underscores.
+ *
+ * @link     https://make.wordpress.org/core/handbook/best-practices/coding-standards/php/#naming-conventions
+ *
+ * @category PHP
+ * @package  PHP_CodeSniffer
+ * @author   Juliette Reinders Folmer <wpplugins_nospam@adviesenzo.nl>
+ */
+class WordPress_Sniffs_NamingConventions_ValidHookNameSniff implements PHP_CodeSniffer_Sniff {
+
+	/**
+	 * Additional word separators.
+	 *
+	 * This public variable allows providing additional word separators which
+	 * will be allowed in hook names via a property in the phpcs.xml config file.
+	 *
+	 * Example usage:
+	 * <rule ref="WordPress.NamingConventions.ValidHookName">
+	 *   <properties>
+	 *     <property name="additionalWordDelimiters" value="-"/>
+	 *   </properties>
+	 * </rule>
+	 *
+	 * Provide several extra delimiters as one string:
+	 * <rule ref="WordPress.NamingConventions.ValidHookName">
+	 *   <properties>
+	 *     <property name="additionalWordDelimiters" value="-/."/>
+	 *   </properties>
+	 * </rule>
+	 *
+	 * @var string
+	 */
+	public $additionalWordDelimiters = '';
+
+	/**
+	 * Regular expression to test for correct punctuation of a hook name.
+	 *
+	 * The placeholder will be replaced by potentially provided additional
+	 * word delimiters in the `prepare_regex()` method.
+	 *
+	 * @var string
+	 */
+	protected $punctuation_regex = '`[^\w%s]`';
+
+	/**
+	 * Functions we're interested in.
+	 *
+	 * Only testing the hook call functions as when using 'add_action'/'add_filter' you can't influence
+	 * the hook name.
+	 *
+	 * @var array
+	 */
+	protected $hook_functions = array(
+		'do_action',
+		'do_action_ref_array',
+		'apply_filters',
+		'apply_filters_ref_array',
+	);
```

Comment:
> We generally make the function names keys so that isset() can be used instead of in_array().

Good point, changing now.

> Also, most of these lists should go in the main sniff class so that they can be reused across sniffs.

I was thinking about that, but then couldn't think of a re-use case. Will move them anyway.

> Also, do_action_deprecated() and apply_filters_deprecated() could be added. https://core.trac.wordpress.org/changeset/37861

Nice one! Hadn't seen that commit yet. Darn! This means I'm going to add that to the `Debug Bar` as well (https://github.com/dd32/debug-bar/pull/17).

For this particular sniff, I'm not sure they should be added. Should deprecated hook names also comply with the naming conventions ? They might have been deprecated because of hook name improvement.
What do you think ?